blog: A Faster Electron (#1138)
electron/electronjs.org-new: blog/a-faster-electron.md

---
title: 'A Faster Electron'

> date: 2026-05-20T00:00:00.000Z

Reviewer: TODO: when this is ready for merge, let's make sure to update the date.

authors: MarshallOfSound
slug: a-faster-electron
tags: [techtalk, internals]
---

import ThemedImage from '@theme/ThemedImage';

> We've spent the last few releases making Electron faster. The work covers startup, IPC, `contextBridge`, networking, module loading, and raw JavaScript throughput, and it applies to every app that runs on Electron. It ships today in Electron 42.3.3, 43.0.0-beta.1, and 44.0.0-nightly.20260603.

Reviewer 2: `43.0.0-beta.1` probably should be updated to `43.0.0` now?

<div>
  <ThemedImage alt="Electron performance improvements. Startup: sandboxed renderer startup drops from about 230 ms to about 130 ms (about 43%); main-process startup drops from about 125 ms to about 75 ms (about 40%). Everything after startup: Speedometer 3.1 on an M5 MacBook rises from 56.6 to 66.2 (about 17%); contextBridge calls are about 28% faster overall." sources={{ light: '/assets/img/blog/faster-startup-hero-light.svg', dark: '/assets/img/blog/faster-startup-hero-dark.svg' }} />
</div>

> The short version: sandboxed renderers start up ~43% faster, the main process boots ~40% faster, and Electron's compiled code got quicker across the board. Speedometer is up ~17%, `contextBridge` calls are up 28-50%, and networking is up 19-40%. You don't have to change a line of your app to get any of this.

Reviewer 3: Assuming you're referring to the Speedometer bench, it reads strangely in the middle of this list of internal framework improvements. (It's not an Electron component, and at the same time devs are more familiar with it so it deserves more of a highlight.)

How about putting it after the other stats as a more explicit punchline e.g.. "Electron now runs the [Speedometer](https://browserbench.org/Speedometer3.1/) web benchmark 17% faster. And you don't have to change a line of your app to get any of this."

---

> title: 'A Faster Electron'

Reviewer: ```suggestion
title: 'Tech Talk: A Faster Electron'
```

If we want to continue the series :)

Author: I'm not 100% convinced this fits into the same group. It serves a primary purpose of saying "hey things got faster" and a secondary purpose of explaining it. I think prefixing with `Tech Talk` diminishes the main point

date: 2026-05-20T00:00:00.000Z
authors: MarshallOfSound
slug: a-faster-electron
tags: [internals]
---

import ThemedImage from '@theme/ThemedImage';

> We've spent the last few releases making Electron faster. The work covers startup, IPC, `contextBridge`, networking, module loading, and raw JavaScript throughput, and it applies to every app that runs on Electron.

Reviewer: > We've spent the last few releases making Electron faster.

We should mention in what versions people can expect to see these changes once above the fold, and again at the end of the post, prob in the "What this means for your app" section.

Links to individual PRs for this work would also be great!

Author: It's a long list of PRs so maybe skip that part, but can definitely include versions here. It's `43.0.0-beta.1` and `42.3.3` and `v44.0.0-nightly.20260603`

<div>
  <ThemedImage alt="Electron performance improvements. Startup: sandboxed renderer startup drops from about 230 ms to about 130 ms (about 43%); browser-process startup drops from about 125 ms to about 75 ms (about 40%). Everything after startup: Speedometer 3.1 on an M5 MacBook rises from 56.6 to 66.2 (about 17%); contextBridge calls are about 28% faster overall." sources={{ light: '/assets/img/blog/faster-startup-hero-light.svg', dark: '/assets/img/blog/faster-startup-hero-dark.svg' }} />
</div>

> The short version: sandboxed renderers start up ~43% faster, the browser process boots ~40% faster, and Electron's compiled code got quicker across the board. Speedometer is up ~17%, `contextBridge` calls are up 28-50%, and networking is up 19-40%. You don't have to change a line of your app to get any of this.

Reviewer: ```suggestion
The short version: sandboxed renderers start up ~43% faster, the main process boots ~40% faster, and Electron's compiled code got quicker across the board. Speedometer is up ~17%, `contextBridge` calls are up 28-50%, and networking is up 19-40%. You don't have to change a line of your app to get any of this.
```

Let's standardize around `main process` since that's what we use for the most part externally.

> This post is in two parts. The first is startup: three changes that shrink the time between launching an app and seeing pixels. The second is everything after startup, and it begins with the discovery that Electron has been shipping with _Chrome's_ compiler optimization data for years.

Reviewer: ```suggestion
This post is in two parts. The first is startup: three changes that shrink the time between launching an app and seeing pixels. The second is everything after, and it begins with the discovery that Electron has been shipping with the wrong compiler optimization data for years.
```

I think this would make a punchier lede.

Author: I don't like "the wrong" phrasing, because it wasn't _entirely_ wrong and that was the gotcha. Like, if it was totally wrong that's a bit awkward, but it was just like ~25% wrong which is a lot harder to catch

<!--truncate-->

## Part one: faster startup

Before any of your code runs, a freshly spawned Electron process loads the binary, initializes Chromium and V8, bootstraps a Node.js environment (where it has one), and parses and compiles Electron's own framework JavaScript. The last two are pure CPU spent turning JavaScript into bytecode, and they happen on _every launch of every process_.

Part one removes that work from the critical path with three independent changes:

> - Pushing sandboxed-renderer startup data over Mojo instead of a synchronous IPC, and caching the preload bytecode.

Reviewer: Not sure if we need to spend any time explaining Mojo, but we should at least link it via Chrome's documentation (e.g.  https://chromium.googlesource.com/chromium/src/+/HEAD/mojo/README.md)

Author: Let's just link it

<div>
  <ThemedImage alt="Before and after of the synchronous preload IPC. Before: the renderer blocks while the main process is busy with other startup work, so the cheap reply isn't delivered until about 80 ms. After: the main process pushes the bundle one-way during navigation setup and the renderer resumes at about 22 ms." sources={{ light: '/assets/img/blog/sync-ipc-light.svg', dark: '/assets/img/blog/sync-ipc-dark.svg' }} />
</div>

The fix was to stop asking. The browser process already knows everything the renderer needs, so it now pushes that data down with the frame-creation parameters over Mojo, and the synchronous message is gone entirely. The preload scripts also gained a bytecode cache, so repeat launches deserialize them instead of re-compiling.

Together these make sandboxed renderer startup roughly 43% faster under real-world conditions, and the renderer's pre-paint time no longer depends on how busy your main process is.

### A build-time code cache for the framework bundles

> Electron's framework JavaScript is embedded in the binary as source, and V8 compiles it from scratch in every process, on every launch. V8 has a standard fix for this, a code cache: compile once, serialize the bytecode, deserialize on later runs. We now generate that cache at build time and embed it next to the source, so no process ever compiles the framework bundles again.

Reviewer: This part should get a bit of rewording to make it a bit clearer that we used to not have the cache and this is a new thing.

### A Node.js startup snapshot for the browser process

A code cache skips _compilation_. The Node.js bootstrap is mostly _execution_: building `process`, wiring the module loader, running ~50 internal setup scripts. Node has a feature designed for exactly this, the startup snapshot: serialize a fully bootstrapped environment once, then deserialize it on every launch instead of re-running the bootstrap. Upstream Node ships with it on. Electron has had it disabled for years.

Why? Electron already boots from two snapshots, and neither is Node's:

1. The V8 startup snapshot: V8's read-only heap and a bare context.
2. The Blink context snapshot: the DOM bindings, with zero compiled JavaScript.
3. The Node snapshot: the bootstrapped Node environment. This is the missing one.

<div>
  <ThemedImage alt="The three snapshot layers stacked: the V8 startup snapshot and the Blink context snapshot were already loaded, but neither captures Node's bootstrap. The Node snapshot, the missing layer, is the one that does." sources={{ light: '/assets/img/blog/snapshot-layers-light.svg', dark: '/assets/img/blog/snapshot-layers-dark.svg' }} />
</div>

> Building the missing one looked impossible at first. Creating a snapshot appears to require a special from-scratch build of V8 that only V8's own tooling gets to use; embedders like Node and Electron only get the "deserialize from an existing snapshot" build, and trying to create a snapshot with it fails with `Heap setup supported only in mksnapshot`.

Reviewer: I'm not a fan of how this section reads. I think if we ask a punchy, direct question like "Why?" we should jump immediately into the answer instead of listing what we already have. Especially with a diagram in the middle, the answer to "why" gets bumped further down.

I think the diagram and the Markdown list also feel redundant.

Measured from process spawn in a release build (the win is everything that happens _before_ your entry script runs), the Node snapshot is worth roughly 40% of browser-process startup, about 50 ms on the hardware tested.

For all three changes, the hard part was the seams rather than the optimizations themselves: Chromium, V8, and Node each have their own model of how a process boots, and the bugs live where those models meet.

Startup is half the story. The other half starts with something we found while looking at how Electron's release builds are compiled.

## Part two: Electron has been shipping with Chrome's optimization data

Modern compilers optimize code around how it actually runs. The biggest lever is Profile-Guided Optimization (PGO): run an instrumented build through real workloads, record which functions are hot, then rebuild with that profile so the compiler knows what to inline, how to lay out branches, and what to keep in the hot path.

> Chrome uses PGO aggressively, and Google publishes fresh Chrome profiles every few hours. Here's the uncomfortable part: Electron's release builds have been applying Chrome's profile. Not a profile of Electron. Chrome's.

Reviewer: The use of "uncomfortable" and "Not a profile of Electron. Chrome's." sounds overly dramatic in tone for this blog post lol

> Chrome uses PGO aggressively, and Google publishes fresh Chrome profiles every few hours. Here's the uncomfortable part: Electron's release builds have been applying Chrome's profile. Not a profile of Electron. Chrome's.

Reviewer: Let's also link to https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/pgo.md

### What borrowing a profile costs

A profile matches functions by name plus a hash of their code. Functions that exist in Electron but not Chrome (all of Node.js, all of Electron's own C++, `contextBridge`) were never in Chrome's profile. Functions that exist in both but are compiled differently in Electron (different patches, flags, V8 configuration) match by name but fail the hash check and are silently rejected. Either way, the compiler gets no guidance and lays the code out as cold.

> We measured it with `llvm-profdata`: about a quarter of the code Electron executes gets zero optimization guidance, and it's concentrated in exactly the code that makes Electron _Electron_.

Reviewer: ```suggestion
We measured it with [`llvm-profdata`](https://llvm.org/docs/CommandGuide/llvm-profdata.html): about a quarter of the code Electron executes gets zero optimization guidance, and it's concentrated in exactly the code that makes Electron _Electron_.
```

<div>
  <ThemedImage alt="What Chrome's profile knows about Electron's code: 74.5% is optimized (name and hash match), 19.3% is silently rejected (same name, different code: Skia SIMD kernels, V8's JSON stringifier, Mojo, net), and 6.2% is not in Chrome's profile at all (all of Node.js, contextBridge, Electron's own C++). One quarter of the code Electron executes gets zero optimization guidance." sources={{ light: '/assets/img/blog/pgo-coverage-light.svg', dark: '/assets/img/blog/pgo-coverage-dark.svg' }} />
</div>

> This isn't theoretical. While doing this work we found that `crypto.randomBytes` in Electron 44 runs at less than half its Electron 42 speed. Nobody touched the crypto code: a BoringSSL patch changed the functions' hashes, Chrome's profile silently stopped covering them, and the compiler started treating them as cold. That's what makes a borrowed profile insidious: code gets slower without anyone changing it, and nothing warns you. With an Electron profile, the regression disappears.

Reviewer: Can we link the patch too? :D

Author: It's not super relevant, just "a patch". I don't think we should link it

> ### Fix one: turn on link-time optimization

Reviewer: I might drop the `fix X:` part of these headings since we don't use them for the first part either

Chromium links with ThinLTO, which lets the compiler optimize across source files at link time, but the default setting does no optimization at all (`--lto-O0`). Chrome's release builds opt into `--lto-O2`. Electron never did.

> Opting in is worth about +5% on Speedometer 3.1 on an M5 MacBook (and more on older hardware). Useful, but as it turns out, the smaller of the two fixes.

Reviewer: I think it'd be good to give a bit more colour on what Speedometer 3.1 is

<div>
  <ThemedImage alt="The cold-marking trap. A profile trained only on browser benchmarks: browser-style code gets 13% faster, but Node.js Buffer operations get 63% slower than stock because the training never ran Node and the profile marked it cold. The enriched profile, trained on Node.js, contextBridge, IPC, TLS networking, and ASAR module loading as well as browser workloads, keeps the browser wins and fully recovers Node." sources={{ light: '/assets/img/blog/cold-marking-light.svg', dark: '/assets/img/blog/cold-marking-dark.svg' }} />
</div>

The training suite now covers what Electron apps actually do: main-process Node.js, `contextBridge` and IPC marshaling, networking over real TLS, module loading from ASAR archives, and compression. It also covers V8's builtins, which have their own separate profile; Chrome's version rejects every promise and async builtin in Electron (we build V8 with promise hooks enabled, Chrome doesn't), so those were running unoptimized too.

### The results

> Each layer stacks on the last. On Speedometer 3.1, on an M5 MacBook:

Reviewer: The blog post mentions benchmarking on an M5 MacBook in various places, but it'd be good to highlight if the PGO optimizations (and others) have the same effects cross-platform.

Author: Fair point, they have similar effects across platforms just we haven't measured them as exactly

| Configuration              | Score | Step |
| -------------------------- | ----- | ---- |

> | Stock Electron             | 56.6  |      |

Reviewer: Is Stock a nightly from before this work landed? I know we mention this in another result down below.

Author: Yeah we can mention this is from a nightly release from before this landed.

| + ThinLTO `--lto-O2`       | 59.2  | +5%  |
| + Electron C++ PGO         | 65.5  | +11% |
| + Electron V8 builtins PGO | 66.2  | +1%  |

> That's +17% end to end, and the biggest single step is Electron's own profile, not the compiler flag. The borrowed optimization data really was the main problem.

Reviewer: ```suggestion
Each layer stacks on the last. On Speedometer 3.1, on an M5 MacBook:

| Configuration              | Score | Step |
| -------------------------- | ----- | ---- |
| Stock Electron             | 56.6  |      |
| + ThinLTO `--lto-O2`       | 59.2  | +5%  |
| + Electron C++ PGO         | 65.5  | +11% |
| + Electron V8 builtins PGO | 66.2  | +1%  |

That's a +17% end-to-end score increase, with the dedicated Electron performance profiles providing most of the results.
```

<div>
  <ThemedImage alt="Speedometer 3.1 on an M5 MacBook as each layer is added: stock Electron 56.6, plus ThinLTO 59.2 (+5%), plus Electron's C++ PGO 65.5 (+11%, the largest step), plus Electron's V8 builtins PGO 66.2. Total +17%, and the biggest gain comes from replacing Chrome's borrowed profile with Electron's own." sources={{ light: '/assets/img/blog/speedometer-stack-light.svg', dark: '/assets/img/blog/speedometer-stack-dark.svg' }} />
</div>

> We also validated the whole stack against the official nightly across a 38-test suite covering `contextBridge`, IPC, and networking, with statistical significance testing: 37 significant improvements, zero regressions, and one tie.

Reviewer: ```suggestion
We also validated the whole stack against the latest [`electron-nightly`](https://www.npmjs.com/package/electron-nightly) build across a 38-test performance suite covering `contextBridge`, IPC, and networking, with statistical significance testing: 37 significant improvements, zero regressions, and one tie.
```

Author: Honestly this whole paragraph can go

| Area                                     | Improvement (geomean) |
| ---------------------------------------- | --------------------- |
| `contextBridge`                          | +28%                  |
| Networking (`fetch`, WebSocket, `https`) | +19%                  |
| IPC                                      | +11%                  |
| Overall                                  | +19.5%                |

> The wins land exactly where Electron's own code runs: `contextBridge` calls that round-trip objects are up 40-55%, `fetch` round-trips are up 23-40%, IPC payloads of every size are up 7-16%. On identical workloads, an optimized Electron now matches or exceeds Chrome itself; the penalty for being "Chrome plus Node.js" instead of Chrome is gone.

Reviewer: > the penalty for being "Chrome plus Node.js"

Was this explained above?

Author: Yes

### What this means for your app

> The same app, on the same hardware, spends less CPU doing what it already does. Chat apps: channel switching and message rendering cost 16-20% less CPU, and every preload API call is 28-50% cheaper. Editors: module loading from ASAR is 8-10% faster and `Buffer`-heavy work is no longer pessimized. Document apps: JSON and structured clone of cached data are 13-37% faster.

Reviewer: ```suggestion
The same app, on the same hardware, spends less CPU doing what it already does. Chat apps: channel switching and message rendering cost 16-20% less CPU, and every preload API call is 28-50% cheaper. Editors: module loading from ASAR is 8-10% faster and `Buffer`-heavy work is optimized. Document apps: JSON and structured clone of cached data are 13-37% faster.
```

This is a smaller note but I've never heard of "pessimized" as an antonym for "optimized" in the performance sense.

Author: I think it's a valid usage here and saying "X is optimized" and "X is no longer pessimized" are not equivalent. One implies better than baseline, the other is "not worse than baseline". Both can be true but the emphasis is "it's no longer slower by default"

> A useful frame: a UI interaction that costs 19.7 ms today misses the 60 fps frame budget and feels janky. At ~19.5% less CPU it costs about 16.4 ms, inside the budget. These changes move real interactions across that line.

Reviewer: what if i have a 144hz gamer monitor :hurtrealbad:

Author: le sigh

---
title: 'A Faster Electron'
date: 2026-05-20T00:00:00.000Z
authors: MarshallOfSound
slug: a-faster-electron

> tags: [internals]

Reviewer: ```suggestion
tags: [techtalk, internals]
```

---

import ThemedImage from '@theme/ThemedImage';

Over the last few releases we've been making Electron faster. Not one feature, and not one benchmark: startup, IPC, `contextBridge`, networking, module loading, and raw JavaScript throughput, across every app that runs on Electron.

<div>

>   <ThemedImage alt="Electron performance improvements. Startup: sandboxed renderer startup drops from about 230 ms to about 130 ms (about 43%); browser-process startup drops from about 125 ms to about 75 ms (about 40%). Everything after startup: Speedometer 3.1 on an M5 MacBook rises from 56.6 to 66.2 (about 17%); contextBridge calls are about 28% faster overall." sources={{ light: '/assets/img/blog/faster-startup-hero-light.svg', dark: '/assets/img/blog/faster-startup-hero-dark.svg' }} />

Reviewer: Is there a public repo with benchmarks that folks can reproduce? Or maybe a repo with instructions on how to run a benchmark on your app?

Author: No these are all ad-hoc benchmarks, do you think they need to be committed somewhere? They're pretty trivial to make on the fly for specific things, I don't want to be maintaining a benchmark repo lol
